### PR TITLE
Module SMPC POC

### LINKED ISSUE #68: Implement nn.Linear with support of MPC
## Feature Description
Based in discussions in #67, we can leverage the spdz protocol (and matrix multiplication) to implement nn.Linear (the basic feed forward block for a neural network) using SMPC

You can check the PyTorch implementation to see how it is implemented in the reference implem.

Don't forget tests ;)

### PR DESCRIPTION
**Pull request by gmuraru**
## Description
Fixes #68 .
Add a Linear layer for SMPC with shares
Use the already implemented SyftModule module from PySyft.

## Affected Dependencies
List any dependencies that are required for this change.

## How has this been tested?
- Added tests for reconstruction
- Added test for inference

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [x] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [x] My changes are covered by tests

### CODE REVIEW

File: src/sympc/protocol/fss/fss.py [hunk 1]
```diff
@@ -215,6 +215,10 @@ def eq(x1, x2):
     def le(x1, x2):
         return fss_op(x1, x2, "comp")
 
+    @staticmethod
+    def relu(x):
+        return x * (x >= 0)
+
```
Comment (LaRiffle) [inline src/sympc/protocol/fss/fss.py:221]: can we move this out of the protocol?

File: src/sympc/store/crypto_primitive_provider.py [hunk 2]
```diff
@@ -53,7 +51,7 @@ def generate_primitives(
     def _transfer_primitives_to_parties(
         op_str: str,
         primitives: List[Any],
-        sessions: List[Session],
+        sessions: List[Any],  # TODO: This should be FIXED to Session
```
Comment (LaRiffle) [inline src/sympc/store/crypto_primitive_provider.py:54]: how can we address this?

File: tests/sympc/module/module_test.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+    clients = get_clients(2)
+
+    session = Session(parties=clients)
+    SessionManager.setup_mpc(session)
+
+    mpc_module = module.share(session=session)
+
+    x = torch.randn(2, 3)
+    x_secret = MPCTensor(secret=x, session=session)
```
Comment (LaRiffle) [inline tests/sympc/module/module_test.py:35]: nit: I think that in other tests x_secret refers to the secret and not to the MPC tensor

File: tests/sympc/module/module_test.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    res = mpc_module.reconstruct()
+
+    assert isinstance(res, sy.Module)
+
+    for name_res, name_expected in zip(res.modules, module.modules):
+        module_expected = module.modules[name_expected]
+
+        assert name_res == name_expected
+
+        module_res = res.modules[name_res]
```
Comment (LaRiffle) [inline tests/sympc/module/module_test.py:68]: `module_res` is defined but not used

File: tests/sympc/module/module_test.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    SessionManager.setup_mpc(session)
+
+    mpc_module = module.share(session=session)
+
+    res = mpc_module.reconstruct()
+
+    assert isinstance(res, sy.Module)
+
+    for name_res, name_expected in zip(res.modules, module.modules):
+        module_expected = module.modules[name_expected]
```
Comment (LaRiffle) [inline tests/sympc/module/module_test.py:64]: `module_expected` is defined but not used

File: src/sympc/module/nn/linear.py [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+from sympc.tensor import MPCTensor
+
+RTOL = 10e-4
+
+
+class Linear:
+    __slots__ = ["weight", "bias", "session", "in_features", "out_features"]
+
+    def __init__(self, session) -> None:
```
Comment (LaRiffle) [inline src/sympc/module/nn/linear.py:20]: So for the moment this class can only be used through a SyNet. Nothing to change, but we need to remember this.

**Review by LaRiffle — CHANGES_REQUESTED** [review]
Great job! We very close from a nice demo :) 
I have left a few questions below

File: src/sympc/module/nn/linear.py (hunk 6 above)
Comment (gmuraru) [reply src/sympc/module/nn/linear.py:20]: Yep

File: src/sympc/store/crypto_primitive_provider.py (hunk 2 above)
Comment (gmuraru) [reply src/sympc/store/crypto_primitive_provider.py:54]: Yes

File: tests/sympc/module/module_test.py (hunk 3 above)
Comment (gmuraru) [reply tests/sympc/module/module_test.py:35]: Done

File: tests/sympc/module/module_test.py (hunk 4 above)
Comment (gmuraru) [reply tests/sympc/module/module_test.py:68]: Used it, forgot to push the changes

File: tests/sympc/module/module_test.py (hunk 5 above)
Comment (gmuraru) [reply tests/sympc/module/module_test.py:64]: Used it, forgot to push the changes

File: src/sympc/module/nn/functional.py [hunk 7]
```diff
@@ -0,0 +1,4 @@
+def relu(x):
+    session = x.session
+    protocol = session.protocol
+    return protocol.relu(x)
```
[the remark this answers is not in the document]
Comment (gmuraru) [reply src/sympc/module/nn/functional.py:4]: Done

File: src/sympc/protocol/fss/fss.py (hunk 1 above)
Comment (gmuraru) [reply src/sympc/protocol/fss/fss.py]: Yep

**Review by LaRiffle — APPROVED** [review]
🙌

**Comment by gmuraru** [comment]
Also fixed #67
